solutions/usaco-918.mdx

```diff
@@ -2,178 +2,168 @@
-		int j = 0;
-		for (int i = 0; i < n; ++i) {
-			while (j + 1 < n && herd[j + 1] - herd[i] < n) {
-				j++;
+		// min is the patch of length n that has the least # of gaps
+		int farthest_cow = 0;
+		for (int curr_cow = 0; curr_cow < n; ++curr_cow) {
+			while (
+				farthest_cow + 1 < n
+				&& herd[farthest_cow + 1] - herd[curr_cow] < n
+			) {
+				farthest_cow++;
 			}
-			minAns = min(minAns, n - (j - i + 1));
+			min_moves = min(min_moves, n - (farthest_cow - curr_cow + 1));
 		}
 	}
-	// Max is just the maximum of the total gap minus either the first or last gap
-	int maxAns = max(tot - (herd[1] - herd[0] - 1), tot - (herd[n - 1] - herd[n - 2] - 1));
-	cout << minAns << "\n" << maxAns;
+
+	// calculate the number of empty cells
+	int gap_num = 0;
+	for (int i = 1; i < n; i++) {
+		gap_num += herd[i] - herd[i - 1] - 1;
+	}
+	// max is the maximum of the total gap minus either the first or last gap
+	int max_moves = max(
+		gap_num - (herd[1] - herd[0] - 1),
+		gap_num - (herd[n - 1] - herd[n - 2] - 1)
+	);
+
+	freopen("herding.out", "w", stdout);
+	cout << min_moves << '\n' << max_moves << endl;
 }
 ```
 
+</CPPSection>
+<JavaSection>
+
 ```java
-//Created by Qi Wang
-import java.util.*;
 import java.io.*;
+import java.util.*;
 
-public class herding {
-	static int N;
-	public static void main(String[] args) throws IOException {
-		InputReader in = new InputReader("herding.in"); //InputReader class down below
-		N = in.nextInt();
-		int tot = 0;
-		int[] pos = new int[N];
-		for (int i = 0; i < N; i++) {
-			pos[i] = in.nextInt();
-		}
-		Arrays.sort(pos);
-
-		int min = Integer.MAX_VALUE;
-		//Getting the total amount of empty cells
-		for (int i = 0; i < N; i++) {
-			if(i > 0)
-				tot += (pos[i]-pos[i-1]-1);
-		}
-		if (pos[N-2]-pos[0] == N-2 && pos[N-1]-pos[N-2]>2 || pos[N-1]-pos[1] == N-2 && pos[1]-pos[0]>2) {
-			min = 2;
-		}
-		else{
-			//Min is the patch that is >= N and has the least number of empty positions
-			int j = 0;
-			for (int i = 0; i < N; i++) {
-				while (j < N-1 && pos[j+1] - pos[i] < N) j++;
-				min = Math.min(min, N-(j-i+1));
-			}
-		}
-
-		//Max is just the maximum of the total gap minus either the first or last gap
-		int max = Math.max(tot - (pos[1] - pos[0] - 1), tot - (pos[pos.length-1] - pos[pos.length-2] - 1));
-		PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("herding.out")));
-		out.println(min);
-		out.println(max);
-		out.close();
-	}
-
-	public static class InputReader {
-		public BufferedReader reader;
-		public StringTokenizer tokenizer;
-
-		public InputReader() {
-			try {
-				reader = new BufferedReader(new InputStreamReader(System.in), 32768);
-			} catch (Exception e) {
-				throw new NullPointerException("Could not create input stream");
-			}
-		}
+public class Herding {
+	public static void main(String[] args) throws IOException{
+		BufferedReader br = new BufferedReader(new FileReader("herding.in"));
```

Comment:
have u closed it?

Reply:
no? shouldn't matter

Reply:
it indeed does not matter, it is just not a good practice leaving resources opened
https://stackoverflow.com/questions/54888904/should-i-always-close-bufferedreader